Authentication to Red Hat Container Registry (Openshift) doesn't work

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

**Description of the issue**:
Using token obtained from Red Hat Container Registry (RHCR) in settings.xml doesn't authenticate user correctly to registry.

**Expected behavior**:
Registry is authenticating user correctly.

**Steps to reproduce**: <!-- Please provide a minimal and precise series of steps -->
mvn compile jib:build with configuration set to point to RHCR and settings.xml containing valid credentials to that. According RHCR username is "unused", amd password is token provided by UI.

**Environment**: <!-- OS, Maven, and other relevant environment information -->
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T22:39:06+03:00)
Java version: 1.8.0_131, vendor: Oracle Corporation
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

jib-maven-plugin version 0.9.6

**`jib-maven-plugin` Configuration**: <!-- Delete this section if not used -->
```xml
<configuration>
	<from>
		<image>openshift-redhat-registry/testProject/test-container</image>
	</from>
	<to>
		<image>openshift-redhat-registry/testProject/${project.artifactId}:${project.version}</image>
	</to>
	<container>
		<jvmFlags>
			<jvmFlag>-Xms512m</jvmFlag>
		</jvmFlags>
	</container>
</configuration>
```

**Log output**: <!-- If applicable, provide relevant log output -->
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:build (default-cli) on project jib-test-project: Build image failed: Tried to retrieve authentication method for openshift-redhat-registry but failed because: Failed get authentication method from 'WWW-Authenticate' header | If this is a bug, please file an issue at https://github.com/GoogleContainerTools/jib/issues/new: Failed to authenticate with the registry because: 'service' was not found in the 'WWW-Authenticate' header, tried to parse: Bearer realm="https://openshift-redhat-registry/openshift/token" -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:build (default-cli) on project jib-test-project: Build image failed
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Build image failed
	at com.google.cloud.tools.jib.maven.BuildImageMojo.execute(BuildImageMojo.java:150)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: com.google.cloud.tools.jib.registry.RegistryErrorException: Tried to retrieve authentication method for openshift-redhat-registry but failed because: Failed get authentication method from 'WWW-Authenticate' header | If this is a bug, please file an issue at https://github.com/GoogleContainerTools/jib/issues/new
	at com.google.cloud.tools.jib.registry.RegistryErrorExceptionBuilder.build(RegistryErrorExceptionBuilder.java:108)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:104)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:31)
	at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:176)
	at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:139)
	at com.google.cloud.tools.jib.registry.RegistryClient.callRegistryEndpoint(RegistryClient.java:338)
	at com.google.cloud.tools.jib.registry.RegistryClient.getRegistryAuthenticator(RegistryClient.java:194)
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator$Initializer.initialize(RegistryAuthenticator.java:83)
	at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call(AuthenticatePushStep.java:85)
	at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call(AuthenticatePushStep.java:42)
	at com.google.common.util.concurrent.CombinedFuture$CallableInterruptibleTask.runInterruptibly(CombinedFuture.java:181)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.cloud.tools.jib.registry.RegistryAuthenticationFailedException: Failed to authenticate with the registry because: 'service' was not found in the 'WWW-Authenticate' header, tried to parse: Bearer realm="https://openshift-redhat-registry/openshift/token"
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.newRegistryAuthenticationFailedException(RegistryAuthenticator.java:150)
	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.fromAuthenticationMethod(RegistryAuthenticator.java:141)
	at com.google.cloud.tools.jib.registry.AuthenticationMethodRetriever.handleHttpResponseException(AuthenticationMethodRetriever.java:98)
	... 13 more
```

**Additional Information**: <!-- Any additional information that may be helpful -->
Failure also occurs when trying:
1. Login with oc CLI tool
2. Using real username and token from `oc whoami -t` as password.

With direct call with Postman to https://openshift-redhat-registry/openshift/token authenticating with user _unused_ and given token, access_token is displayed correctly in response

With real username + that `oc whoami` -t -token that /token -endpoint returns 401.

<!-- Thanks for contributing! -->

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/AuthenticationMethodRetriever.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/AuthenticationMethodRetriever.java
@@ -96,7 +96,7 @@ class AuthenticationMethodRetriever implements RegistryEndpointProvider<Registry
     // Parses the header to retrieve the components.
     try {
       return RegistryAuthenticator.fromAuthenticationMethod(
-          authenticationMethod, registryEndpointRequestProperties.getImageName());
+          authenticationMethod, registryEndpointRequestProperties);
 
     } catch (RegistryAuthenticationFailedException ex) {
       throw new RegistryErrorExceptionBuilder(getActionDescription(), ex)

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
@@ -137,10 +137,11 @@ public class RegistryAuthenticator {
 
     Pattern servicePattern = Pattern.compile("service=\"(.*?)\"");
     Matcher serviceMatcher = servicePattern.matcher(authenticationMethod);
-    if (!serviceMatcher.find()) {
-      throw newRegistryAuthenticationFailedException(authenticationMethod, "service");
+    // use the provided registry as the service (e.g., for OpenShift)
+    String service = repository;
+    if (serviceMatcher.find()) {
+      service = serviceMatcher.group(1);
     }
-    String service = serviceMatcher.group(1);
 
     return new RegistryAuthenticator(realm, service, repository);
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
@@ -108,6 +108,7 @@ public class RegistryAuthenticator {
    * Instantiates from parsing a {@code WWW-Authenticate} header.
    *
    * @param authenticationMethod the {@code WWW-Authenticate} header value
+   * @param registryEndpointRequestProperties the registry request properties
    * @param repository the repository/image name
    * @return a new {@link RegistryAuthenticator} for authenticating with the registry service
    * @throws RegistryAuthenticationFailedException if authentication fails
@@ -116,7 +117,9 @@ public class RegistryAuthenticator {
    */
   @Nullable
   static RegistryAuthenticator fromAuthenticationMethod(
-      String authenticationMethod, String repository) throws RegistryAuthenticationFailedException {
+      String authenticationMethod,
+      RegistryEndpointRequestProperties registryEndpointRequestProperties)
+      throws RegistryAuthenticationFailedException {
     // If the authentication method starts with 'basic ' (case insensitive), no registry
     // authentication is needed.
     if (authenticationMethod.matches("^(?i)(basic) .*")) {
@@ -137,10 +140,14 @@ public class RegistryAuthenticator {
 
     Pattern servicePattern = Pattern.compile("service=\"(.*?)\"");
     Matcher serviceMatcher = servicePattern.matcher(authenticationMethod);
-    // use the provided registry as the service when missing (e.g., for OpenShift)
-    String service = serviceMatcher.find() ? serviceMatcher.group(1) : repository;
+    // use the provided registry location when missing service (e.g., for OpenShift)
+    String service =
+        serviceMatcher.find()
+            ? serviceMatcher.group(1)
+            : registryEndpointRequestProperties.getServerUrl();
 
-    return new RegistryAuthenticator(realm, service, repository);
+    return new RegistryAuthenticator(
+        realm, service, registryEndpointRequestProperties.getImageName());
   }
 
   private static RegistryAuthenticationFailedException newRegistryAuthenticationFailedException(

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java
@@ -137,11 +137,8 @@ public class RegistryAuthenticator {
 
     Pattern servicePattern = Pattern.compile("service=\"(.*?)\"");
     Matcher serviceMatcher = servicePattern.matcher(authenticationMethod);
-    // use the provided registry as the service (e.g., for OpenShift)
-    String service = repository;
-    if (serviceMatcher.find()) {
-      service = serviceMatcher.group(1);
-    }
+    // use the provided registry as the service when missing (e.g., for OpenShift)
+    String service = serviceMatcher.find() ? serviceMatcher.group(1) : repository;
 
     return new RegistryAuthenticator(realm, service, repository);
   }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStepTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStepTest.java
@@ -59,7 +59,7 @@ public class BuildAndCacheApplicationLayerStepTest {
   private Path temporaryCacheDirectory;
 
   @Before
-  public void setUp() throws IOException, URISyntaxException {
+  public void setUp() throws IOException {
     Mockito.when(mockBuildConfiguration.getBuildLogger()).thenReturn(new TestBuildLogger());
     temporaryCacheDirectory = temporaryFolder.newFolder().toPath();
   }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryAuthenticatorTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryAuthenticatorTest.java
@@ -23,6 +23,8 @@ import org.junit.Test;
 
 /** Tests for {@link RegistryAuthenticator}. */
 public class RegistryAuthenticatorTest {
+  private final RegistryEndpointRequestProperties registryEndpointRequestProperties =
+      new RegistryEndpointRequestProperties("someserver", "someimage");
 
   @Test
   public void testFromAuthenticationMethod_bearer()
@@ -30,7 +32,7 @@ public class RegistryAuthenticatorTest {
     RegistryAuthenticator registryAuthenticator =
         RegistryAuthenticator.fromAuthenticationMethod(
             "Bearer realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
-            "someimage");
+            registryEndpointRequestProperties);
     Assert.assertEquals(
         new URL("https://somerealm?service=someservice&scope=repository:someimage:scope"),
         registryAuthenticator.getAuthenticationUrl("scope"));
@@ -38,7 +40,7 @@ public class RegistryAuthenticatorTest {
     registryAuthenticator =
         RegistryAuthenticator.fromAuthenticationMethod(
             "bEaReR realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
-            "someimage");
+            registryEndpointRequestProperties);
     Assert.assertEquals(
         new URL("https://somerealm?service=someservice&scope=repository:someimage:scope"),
         registryAuthenticator.getAuthenticationUrl("scope"));
@@ -49,24 +51,25 @@ public class RegistryAuthenticatorTest {
     Assert.assertNull(
         RegistryAuthenticator.fromAuthenticationMethod(
             "Basic realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
-            "someimage"));
+            registryEndpointRequestProperties));
 
     Assert.assertNull(
         RegistryAuthenticator.fromAuthenticationMethod(
             "BASIC realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
-            "someimage"));
+            registryEndpointRequestProperties));
 
     Assert.assertNull(
         RegistryAuthenticator.fromAuthenticationMethod(
             "bASIC realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
-            "someimage"));
+            registryEndpointRequestProperties));
   }
 
   @Test
   public void testFromAuthenticationMethod_noBearer() {
     try {
       RegistryAuthenticator.fromAuthenticationMethod(
-          "realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"", "someimage");
+          "realm=\"https://somerealm\",service=\"someservice\",scope=\"somescope\"",
+          registryEndpointRequestProperties);
       Assert.fail("Authentication method without 'Bearer ' or 'Basic ' should fail");
 
     } catch (RegistryAuthenticationFailedException ex) {
@@ -79,7 +82,8 @@ public class RegistryAuthenticatorTest {
   @Test
   public void testFromAuthenticationMethod_noRealm() {
     try {
-      RegistryAuthenticator.fromAuthenticationMethod("Bearer scope=\"somescope\"", "someimage");
+      RegistryAuthenticator.fromAuthenticationMethod(
+          "Bearer scope=\"somescope\"", registryEndpointRequestProperties);
       Assert.fail("Authentication method without 'realm' should fail");
 
     } catch (RegistryAuthenticationFailedException ex) {
@@ -94,10 +98,10 @@ public class RegistryAuthenticatorTest {
       throws MalformedURLException, RegistryAuthenticationFailedException {
     RegistryAuthenticator registryAuthenticator =
         RegistryAuthenticator.fromAuthenticationMethod(
-            "Bearer realm=\"https://somerealm\"", "someimage");
+            "Bearer realm=\"https://somerealm\"", registryEndpointRequestProperties);
 
     Assert.assertEquals(
-        new URL("https://somerealm?service=someimage&scope=repository:someimage:scope"),
+        new URL("https://somerealm?service=someserver&scope=repository:someimage:scope"),
         registryAuthenticator.getAuthenticationUrl("scope"));
   }
 }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryAuthenticatorTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryAuthenticatorTest.java
@@ -90,16 +90,14 @@ public class RegistryAuthenticatorTest {
   }
 
   @Test
-  public void testFromAuthenticationMethod_noService() {
-    try {
-      RegistryAuthenticator.fromAuthenticationMethod(
-          "Bearer realm=\"https://somerealm\"", "someimage");
-      Assert.fail("Authentication method without 'service' should fail");
+  public void testFromAuthenticationMethod_noService()
+      throws MalformedURLException, RegistryAuthenticationFailedException {
+    RegistryAuthenticator registryAuthenticator =
+        RegistryAuthenticator.fromAuthenticationMethod(
+            "Bearer realm=\"https://somerealm\"", "someimage");
 
-    } catch (RegistryAuthenticationFailedException ex) {
-      Assert.assertEquals(
-          "Failed to authenticate with the registry because: 'service' was not found in the 'WWW-Authenticate' header, tried to parse: Bearer realm=\"https://somerealm\"",
-          ex.getMessage());
-    }
+    Assert.assertEquals(
+        new URL("https://somerealm?service=someimage&scope=repository:someimage:scope"),
+        registryAuthenticator.getAuthenticationUrl("scope"));
   }
 }

